daemon/gluster-blockd.c

```diff
@@ -366,22 +352,114 @@ gbMinKernelVersionCheck(void)
       distro, tptr, verStr.release);
 
  fail:
-  pclose(fp);
+  GB_FREE(distro);
 
   exit(EXIT_FAILURE);
 }
 
 
+static bool
+gbDependencyVersionCompare(int dependencyName, char *version)
+{
+  size_t vNum[VERNUM_BUFLEN] = {0, };
+  char *verStr;
+  int i = 0, j;
+  char *token, *tmp;
+  bool done = false;
+  bool ret = false;
+
+
+  if (GB_STRDUP(verStr, version) < 0) {
+    LOG("mgmt", GB_LOG_ERROR,
+        "gbDependencyVersionCompare: failed to strdup (%s)", strerror(errno));
+    return ret;
+  }
+
+  token = strtok(verStr, ".-");
+  while( token != NULL ) {
+    done = false;
+    for (j = 0; j < strlen(token); j++) { /* say if version is 2.1.fb49, parse fb49 */
+      if (isdigit(token[j])) {
+        vNum[i] = atoi(token);
+      } else {
+        tmp = token;
+        for(; *tmp; ++tmp) {
+          if (isdigit(*tmp)) {
+            vNum[i] = atoi(tmp);  /* feed 49 from fb49, so done = true */
+            done = true;
+            break;
+          }
+        }
+      }
+      if (done) {
+        break;
+      }
+    }
+    token = strtok(NULL, ".-");
+    i++;
+  }
+
+  switch (dependencyName) {
+  case TCMURUNNER:
+    if (DEPENDENCY_VERSION(vNum[0], vNum[1], vNum[2]) >= GB_MIN_TCMURUNNER_VERSION_CODE) {
+        ret = true;
+    }
+    break;
+  case TARGETCLI:
+    if (DEPENDENCY_VERSION(vNum[0], vNum[1], vNum[2]) >= GB_MIN_TARGETCLI_VERSION_CODE) {
+        ret = true;
+    }
+    break;
+  }
+
+  GB_FREE(verStr);
+  return ret;
+}
+
+
+static void
+gbDependenciesVersionCheck(void)
+{
+  char *out = NULL;
+
+
+  out = gbRunnerGetOutput("tcmu-runner --version 2>&1 | awk -F' ' '{printf $NF}'");
+  if (!gbDependencyVersionCompare(TCMURUNNER, out)) {
+    LOG ("mgmt", GB_LOG_ERROR,
+         "current tcmu-runner version is %s, gluster-block need atleast - %s",
+         out, GB_MIN_TCMURUNNER_VERSION);
+    goto out;
+  }
+  LOG("mgmt", GB_LOG_INFO, "starting with tcmu-runner version - %s", out);
+  GB_FREE(out);
+
+  out = gbRunnerGetOutput("targetcli --version 2>&1 | awk -F' ' '{printf $NF}'");
+  if (!gbDependencyVersionCompare(TARGETCLI, out)) {
+    LOG ("mgmt", GB_LOG_ERROR,
+         "current targetcli version is %s, gluster-block need atleast - %s",
+         out, GB_MIN_TARGETCLI_VERSION);
+    goto out;
+  }
+  LOG("mgmt", GB_LOG_INFO, "starting with targetcli version - %s", out);
+  GB_FREE(out);
+
+  return;
+
+ out:
+  GB_FREE(out);
+  LOG("mgmt", GB_LOG_ERROR, "%s",
+      "Please install the recommended dependency version and try again.");
+  exit(EXIT_FAILURE);
+}
+
+
```

Comment:
While all the above looks fine, it makes me sad that we need to write so much code in C to do it :-/

No other option though.

Reply:
Indeed :-(